Mathlib/Analysis/Convex/Cone/Proper.lean

```diff
@@ -30,77 +32,107 @@ The next steps are:
-    [TopologicalSpace E] [Module 𝕜 E] extends Submodule {c : 𝕜 // 0 ≤ c} E where
-  isClosed' : IsClosed (carrier : Set E)
+abbrev ProperCone := ClosedSubmodule R≥0 E
 
 namespace ProperCone
 section Module
+variable {C C₁ C₂ : ProperCone R E} {r : R} {x : E}
 
-variable {𝕜 : Type*} [Semiring 𝕜] [PartialOrder 𝕜] [IsOrderedRing 𝕜]
-variable {E : Type*} [AddCommMonoid E] [TopologicalSpace E] [Module 𝕜 E]
+/-- Alias of `ClosedSubmodule.toSubmodule` for convenience and discoverability. -/
+@[coe] abbrev toPointedCone (C : ProperCone R E) : PointedCone R E := C.toSubmodule
 
-/-- A `PointedCone` is defined as an alias of submodule. We replicate the abbreviation here and
-define `toPointedCone` as an alias of `toSubmodule`. -/
-abbrev toPointedCone (C : ProperCone 𝕜 E) := C.toSubmodule
+instance : Coe (ProperCone R E) (PointedCone R E) := ⟨toPointedCone⟩
 
-attribute [coe] toPointedCone
+lemma toPointedCone_injective : Injective ((↑) : ProperCone R E → PointedCone R E) :=
+  ClosedSubmodule.toSubmodule_injective
 
-instance : Coe (ProperCone 𝕜 E) (PointedCone 𝕜 E) :=
-  ⟨toPointedCone⟩
+-- TODO: add `ConvexConeClass` that extends `SetLike` and replace the below instance
+instance : SetLike (ProperCone R E) E where
+  coe C := C.carrier
+  coe_injective' _ _ h := ProperCone.toPointedCone_injective <| SetLike.coe_injective h
 
-theorem toPointedCone_injective : Function.Injective ((↑) : ProperCone 𝕜 E → PointedCone 𝕜 E) :=
-  fun S T h => by cases S; cases T; congr
+@[ext] lemma ext (h : ∀ x, x ∈ C₁ ↔ x ∈ C₂) : C₁ = C₂ := SetLike.ext h
 
--- TODO: add `ConvexConeClass` that extends `SetLike` and replace the below instance
-instance : SetLike (ProperCone 𝕜 E) E where
-  coe K := K.carrier
-  coe_injective' _ _ h := ProperCone.toPointedCone_injective (SetLike.coe_injective h)
+@[simp] lemma mem_toPointedCone : x ∈ C.toPointedCone ↔ x ∈ C := .rfl
 
-@[ext]
-theorem ext {S T : ProperCone 𝕜 E} (h : ∀ x, x ∈ S ↔ x ∈ T) : S = T :=
-  SetLike.ext h
+protected lemma pointed_toConvexCone (C : ProperCone R E) : (C : ConvexCone R E).Pointed :=
+  C.toPointedCone.pointed_toConvexCone
 
-@[simp]
-theorem mem_coe {x : E} {K : ProperCone 𝕜 E} : x ∈ (K : PointedCone 𝕜 E) ↔ x ∈ K :=
-  Iff.rfl
+protected lemma nonempty (C : ProperCone R E) : (C : Set E).Nonempty := C.toSubmodule.nonempty
+protected lemma isClosed (C : ProperCone R E) : IsClosed (C : Set E) := C.isClosed'
 
-instance instZero (K : ProperCone 𝕜 E) : Zero K := PointedCone.instZero (K.toSubmodule)
+protected nonrec lemma smul_mem (C : ProperCone R E) (hx : x ∈ C) (hr : 0 ≤ r) : r • x ∈ C :=
+  C.smul_mem ⟨r, hr⟩ hx
+
+section T1Space
+variable [T1Space E]
+
+lemma mem_bot : x ∈ (⊥ : ProperCone R E) ↔ x = 0 := .rfl
+
+@[simp, norm_cast] lemma toPointedCone_bot : (⊥ : ProperCone R E).toPointedCone = ⊥ := rfl
+
+end T1Space
+
+/-- The closure of image of a proper cone under a `ℝ`-linear map is a proper cone. We
+use continuous maps here so that the comap of f is also a map between proper cones. -/
```

Comment:
If you didn't insist `f` was continuous, what do you get out instead? I presume just a convex cone or something? (Not an objection, just out of curiosity)